File: 프로그래머스/lv1/최원준/크레인 인형뽑기 게임.py
def solution(board, moves):
    # print(*board, sep='\n')
    n = len(board)
    m = len(board[0])
    cords = []
    for col in range(m):
        found=False
        for row in range(n):
            if board[row][col] != 0:
                cords.append(row)
                break
        else:
            cords.append(-1)

    ans = 0
    stack = []
    for idx in moves:
        idx-=1
        if cords[idx] == -1: continue
        row = cords[idx]
        val = board[row][idx]
        if stack and stack[-1] == val:
            stack.pop()
            ans+=2
        else:
            stack.append(val)
        if row+1 == n:
            cords[idx] = -1
        else:
            cords[idx] = row+1
    return ans

File: 프로그래머스/lv1/김원래/크레인인형뽑기게임.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

stack<int> basket;
int topRows[31];
int lastRow;

int solution(vector<vector<int>> board, vector<int> moves) {
    int answer = 0;
    lastRow = board.size();

    for (int col = 0; col < board[0].size(); ++col)
    {
        int curRow = 0;
        while (board[curRow][col] == 0) ++curRow;
        topRows[col] = curRow;
    }


    for (int move = 0; move < moves.size(); ++move)
    {

        int col = moves[move] - 1;

        if (topRows[col] == lastRow) continue;

        if (!basket.empty() && (board[topRows[col]][col] == basket.top()))
        {
            basket.pop();
            answer += 2;
        }
        else
            basket.push(board[topRows[col]][col]);


        topRows[col]++;
    }

    return answer;
}